go/worker/sandbox/containerdPool.go

```diff
@@ -0,0 +1,311 @@
+        labels        map[string]string
+        idxPtr        *int64
+        eventHandlers []SandboxEventFunc
+        debugger
+}
+
+// NewContainerdPool creates a ContainerdPool.
+func NewContainerdPool() (*ContainerdPool, error) {
+        client, err := containerd.New(common.Conf.Containerd.SocketAddress,
+                containerd.WithDefaultNamespace(common.Conf.Containerd.Namespace),
+                containerd.WithTimeout(10*time.Second),
+        )
+        if err != nil {
+                return nil, fmt.Errorf("failed to connect to containerd: %v", err)
+        }
+
+        ctx := namespaces.WithNamespace(context.Background(), common.Conf.Containerd.Namespace)
+
+        // ensure that containerd is serving
+        healthCtx, cancel := context.WithTimeout(ctx, 5*time.Second)
+        defer cancel()
+        serving, err := client.IsServing(healthCtx)
+        if err != nil {
+                return nil, fmt.Errorf("failed to check containerd health: %v", err)
+        }
+        if !serving {
+                return nil, fmt.Errorf("containerd is not serving at %s", common.Conf.Containerd.SocketAddress)
+        }
+        slog.Info("Successfully connected to containerd", "socket_address", common.Conf.Containerd.SocketAddress, "namespace", common.Conf.Containerd.Namespace)
+        var sharedIdx int64 = -1
+        pool := &ContainerdPool{
+                client:    client,
+                namespace: common.Conf.Containerd.Namespace,
+                runtime:   common.Conf.Containerd.Runtime,
+                image:     common.Conf.Containerd.Base_image,
+                labels: map[string]string{
+                        containerdutil.CONTAINERD_LABEL_CLUSTER: common.Conf.Worker_dir,
+                        containerdutil.CONTAINERD_LABEL_RUNTIME: common.Conf.Containerd.Runtime,
+                },
+                idxPtr:        &sharedIdx,
+                eventHandlers: []SandboxEventFunc{},
+        }
+
+        pool.debugger = newDebugger(pool)
+
+        return pool, nil
+
+}
+
+
+func (pool *ContainerdPool) Create(parent Sandbox, isLeaf bool, codeDir, scratchDir string, meta *SandboxMeta, rtType common.RuntimeType) (sb Sandbox, err error) {
+
+        socketPath := filepath.Join(scratchDir, "ol.sock")
+        if len(socketPath) > 108 { // check ealier to avoid resource cleanup
+                return nil, fmt.Errorf("socket path length cannot exceed 108 characters (try moving cluster closer to the root directory")
+        }
+        
+        meta = fillMetaDefaults(meta)
+        if parent != nil || !isLeaf {
+                return nil, fmt.Errorf("currently ContainerdPool does not support forking from a parent, so parent must be nil and isLeaf must be true")
+        }
+        id := fmt.Sprintf("ctrd-%d", atomic.AddInt64(pool.idxPtr, 1))
+
+        ctx := namespaces.WithNamespace(context.Background(), pool.namespace)
+        if err := containerdutil.EnsureImageExists(ctx, pool.client, pool.image); err != nil {
```

Comment:
I wonder if it is expensive to check the image every time we create a container?  Would it make sense to do it during pool creation instead of container creation?